README.md
Qi Daozei
=========

Qi Daozei (QDZ) is an Oriental-themed fantasy roguelike game for T-Engine4,
the same engine that powers ToME.

For more information about ToME and T-Engine4, visit http://te4.org.


CHANGELOG.md
Version 0.4.1, "An Apropos Word" (2013-12-16)
=============================================

* Noteworthy UI improvements
    * Character sheets and map tooltips include non-physical melee damage (such as a fire ant's bonus fire damage).
    * When inspecting an enemy, the character sheet's tooltips for the enemy's Attack and Defense values show the hit chances for you versus the enemy.
    * Character sheets list special movement qualities (flying, random movement, etc.).
    * Proficiencies' tooltips provide more information about proficiency level.
    * Digging implements' tooltips tell how long digging will take.
* Gameplay tweaks
    * Boost Fourth Blessing: Longevity.
    * You can now use the Mining talent without a digging implement (although it will take longer).
    * Healing potions' effects now scale with constitution.
* Other UI improvements
    * Improve object tooltips for multiple objects.
    * Improved Pick Up Items dialog.
* Bugfixes
    * Inventory screen tooltips sometimes failed to show.
    * A missed Fire Slash failed to display its "near miss" message.
    * Saving a game while Poisoned Dart was in flight didn't work properly.
* Minor changes
    * Adjust the effect of Mining proficiency on Poison Ore Strike.
    * Reading scrolls now requires the ability to see.
* Compatible with T-Engine 1.1.0

Version 0.4.0, "Desperate Stratagems" (2013-11-30)
==================================================

Major changes:

* Boot screen
* Critical hits.
    * Each attack has a 5% chance of knocking an opponent off balance.
    * The next attack against an off-balance opponent is an automatic critical hit. In addition to dealing 50% bonus damage, critical hits allow you to execute weapon-specific maneuvers, such as a staff sweep or spinning halberd attack.
* Desperation attacks. When reduced to 30% health or lower, your next attack is an automatic critical hit.
* Changes to Off-Hand Attack make learning left hand techniques easier:
    * Staffs are now classified as double weapons and so can be used with Off-Hand Attack.  (Currently, double weapons _only_ help with Off-Hand Attack; rumors exist of fighting styles that take full advantage of double weapons, but such techniques cannot be found this early in the dungeon...)
    * Off-Hand Attack's tooltip now gives more information, to help you know whether or not it can deal the finishing blow.

Less visible changes:

* Rebalance health, damage, and armor.
* Various tweaks to techniques:
    * Blood Sip heals on hit as well as on each turn.
    * You can now step past unconscious opponents who you've chosen not to kill.
    * Pickpocketing an unconscious creature always succeeds.
    * Creatures who've regained consciousness are still groggy and unable to act for one turn.
    * Poisoned Dart's poison is more dangerous.
    * Charged Bolt is more accurate.
    * Increase damage on Explosive Tags.
* Various tweaks to creatures.  In particular:
    * Fire ants are rarer.
    * Cave bats know Blood Sip.

Version 0.3.0, "A Warrior's Tools" (2013-11-18)
===============================================

* New items: meditation scrolls, scrolls of body hardening, smoke bombs, and explosive tags. Explosive tags are particularly fun to play with.
* AI fixes: dog-head men no longer shoot each other in the back, fire ants no longer try to slash you from across the room, and so on.
* Various other improvements, such as a random name generator and updated particle effects
* Compatible with T-Engine 1.0.5

Version 0.2.0 (2013-10-12)
==========================

* 3 new monsters to fight
* New monsters mean new qi techniques to absorb. Pursue happiness and good fortune with the Five Blessings, lurk in the shadows with Dweller in Darkness and Bat-Blind Vision, amass a swarm of insect allies with Hive Mind, and more.
* New game mechanics: Blindsense and blind-fighting now exist. (They're currently only useful when it's dark; blindness itself is coming in a future version.) You can pickpocket your enemies. There's now a limit to how many qi techniques you can know at once, so absorbing a new technique may require that you forget an old one.

class/Game.lua
-- Qi Daozei
-- Copyright (C) 2013 Josh Kelley
--
-- based on
-- ToME - Tales of Middle-Earth
-- Copyright (C) 2009, 2010, 2011, 2012, 2013 Nicolas Casalini
--
-- This program is free software: you can redistribute it and/or modify
-- it under the terms of the GNU General Public License as published by
-- the Free Software Foundation, either version 3 of the License, or
-- (at your option) any later version.
--
-- This program is distributed in the hope that it will be useful,
-- but WITHOUT ANY WARRANTY; without even the implied warranty of
-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-- GNU General Public License for more details.
--
-- You should have received a copy of the GNU General Public License
-- along with this program.  If not, see <http://www.gnu.org/licenses/>.

require "engine.class"
require "engine.GameTurnBased"
require "engine.interface.GameTargeting"
require "engine.KeyBind"
local Savefile = require "engine.Savefile"
local DamageType = require "engine.DamageType"
local Zone = require "engine.Zone"
local Map = require "engine.Map"
local Level = require "engine.Level"
local Birther = require "mod.dialogs.Birther"
local Dialog = require "mod.class.ui.Dialog"

local Grid = require "mod.class.Grid"
local Actor = require "mod.class.Actor"
local Player = require "mod.class.Player"
local NPC = require "mod.class.NPC"
local PlayerDisplay = require "mod.class.PlayerDisplay"

local HotkeysDisplay = require "engine.HotkeysDisplay"
local ActorsSeenDisplay = require "engine.ActorsSeenDisplay"
local LogDisplay = require "engine.LogDisplay"
local LogFlasher = require "engine.LogFlasher"
local FlyingText = require "engine.FlyingText"
local Tooltip = require "engine.Tooltip"
local GameUI = require "mod.class.ui.GameUI"

local MapMenu = require "mod.dialogs.MapMenu"

module(..., package.seeall, class.inherit(engine.GameTurnBased, engine.interface.GameTargeting))

function _M:init()
    engine.GameTurnBased.init(self, engine.KeyBind.new(), 1000, 100)

    -- Pause at birth
    self.paused = true

    -- Same init as when loaded from a savefile
    self:loaded()
end

function _M:run()
    local player_display_width = 200
    self.flash = LogFlasher.new(player_display_width + 8, 0, self.w, 20, nil, nil, nil, {255,255,255}, {30,30,30})
    self.logdisplay = LogDisplay.new(0, self.h * 0.8, self.w * 0.5, self.h * 0.2, nil, GameUI.font_name, GameUI.font_size, {255,255,255}, {30,30,30})
    self.hotkeys_display = HotkeysDisplay.new(nil, self.w * 0.5, self.h * 0.8, self.w * 0.5, self.h * 0.2, {30,30,0}, GameUI.mono_font_name, GameUI.font_size - 2)
    self.npcs_display = ActorsSeenDisplay.new(nil, self.w * 0.5, self.h * 0.8, self.w * 0.5, self.h * 0.2, {30,30,0})
    self.tooltip = Tooltip.new(GameUI.font_name, GameUI.font_size, {255,255,255}, {30,30,30})
    self.player_display = PlayerDisplay.new(0, 0, player_display_width, self.h * 0.8, {0,0,0}, GameUI.mono_font_name, GameUI.font_size)
    self.flyers = FlyingText.new()
    self:setFlyingText(self.flyers)

    local function logSee(e)
        return e and (self.level.map.seens(e.x, e.y) or e == game.player)
    end

    self.log = function(style, ...) if type(style) == "number" then self.logdisplay(...) self.flash(style, ...) else self.logdisplay(style, ...) self.flash(self.flash.NEUTRAL, style, ...) end end
    self.logSeen = function(e, style, ...) if logSee(e) then self.log(style, ...); return true end end
    self.logSeenAny = function(e, style, ...) for i, v in pairs(e) do if logSee(v) then self.log(style, ...); return true end end end
    self.logPlayer = function(e, style, ...) if e == self.player then self.log(style, ...) end end

    self.logAnySeen = function(e_list, style, ...)
        for i, e in ipairs(e_list) do
            if self.logSeen(e, style, ...) then return true end
        end
    end

    self.log(self.flash.GOOD, "Welcome to #00FF00#Qi Daozei!")

    -- Setup inputs
    self:setupCommands()
    self:setupMouse()

    -- Starting from here we create a new game
    if not self.player then self:newGame() end

    self.hotkeys_display.actor = self.player
    self.npcs_display.actor = self.player

    -- Setup the targeting system
    engine.interface.GameTargeting.init(self)

    -- OK, everything is good to go, activate the game in the engine!
    self:setCurrent()

    if self.level then self:setupDisplayMode() end
end

function _M:newGame()
    self.player = Player.new{name=self.player_name, game_ender=true}
    Map:setViewerActor(self.player)
    self:setupDisplayMode()

    self.creating_player = true
    local birth = Birther.new(nil, self.player, {"base", "sex", "background"}, function()
        self:changeLevel(1, "dungeon")
        print("[PLAYER BIRTH] resolve...")
        self.player:resolve()
        self.player:resolve(nil, true)
        self.player.energy.value = self.energy_to_act
        self.paused = true
        self.creating_player = false
        print("[PLAYER BIRTH] resolved!")

        self.player:sortHotkeysByTalent()
        self.player:sortHotkeys()

        -- HACK: This forces PlayerDisplay and HotkeysDisplay to update after birth descriptors are finished.
        self.player.changed = true
    end)
    self:registerDialog(birth)
end

function _M:loaded()
    engine.GameTurnBased.loaded(self)
    Zone:setup{npc_class="mod.class.NPC", grid_class="mod.class.Grid", object_class="mod.class.Object"}
    Map:setViewerActor(self.player)
    Map:setViewPort(200, 20, self.w - 200, math.floor(self.h * 0.80) - 20, 32, 32, nil, 22, true)
    self.key = engine.KeyBind.new()
end

function _M:setupDisplayMode()
    print("[DISPLAY MODE] 32x32 ASCII/background")
    Map:setViewPort(200, 20, self.w - 200, math.floor(self.h * 0.80) - 20, 32, 32, nil, 22, true)
    Map:resetTiles()
    Map.tiles.use_images = false

    if self.level then
        self.level.map:recreate()
        engine.interface.GameTargeting.init(self)
        self.level.map:moveViewSurround(self.player.x, self.player.y, 8, 8)
    end
end

function _M:save()
    return class.save(self, self:defaultSavedFields{}, true)
end

function _M:getSaveDescription()
    return {
        name = self.player.name,
        description = ([[Exploring level %d of %s.]]):format(self.level.level, self.zone.name),
    }
end

function _M:leaveLevel(level, lev, old_lev)
    if level:hasEntity(self.player) then
        level.exited = level.exited or {}
        if lev > old_lev then
            level.exited.down = {x=self.player.x, y=self.player.y}
        else
            level.exited.up = {x=self.player.x, y=self.player.y}
        end
        level.last_turn = game.turn
        level:removeEntity(self.player)
    end
end

function _M:changeLevel(lev, zone)
    local old_lev = (self.level and not zone) and self.level.level or -1000
    if zone then
        if self.zone then
            self.zone:leaveLevel(false, lev, old_lev)
            self.zone:leave()
        end
        if type(zone) == "string" then
            self.zone = Zone.new(zone)
        else
            self.zone = zone
        end
    end
    self.zone:getLevel(self, lev, old_lev)

    if lev > old_lev then
        self.player:move(self.level.default_up.x, self.level.default_up.y, true)
    else
        self.player:move(self.level.default_down.x, self.level.default_down.y, true)
    end
    self.level:addEntity(self.player)

    self.level.delayed_crit = {}
end

function _M:getPlayer()
    return self.player
end

--- Says if this savefile is usable or not
function _M:isLoadable()
    return not self:getPlayer(true).dead
end

function _M:tick()
    if self.level then
        self:targetOnTick()

        engine.GameTurnBased.tick(self)
    else
        engine.Game.tick(self)
    end
    -- When paused (waiting for player input) we return true: this means we wont be called again until an event wakes us
    if self.paused and not savefile_pipe.saving then return true end
end

--- Called every game turns
-- Does nothing, you can override it
function _M:onTurn()
    self:processCrit()

    -- The following happens only every 10 game turns (once for every turn of 1 mod speed actors)
    if self.turn % 10 ~= 0 then return end

    -- Process overlay effects
    self.level.map:processEffects()
end

---Adds e to the list of entities / actors that have been knocked off balance
-- to receive a critical hit.
function _M:addCrit(e)
    -- If e is already scheduled for a delayed crit, apply the crit now
    -- (*and* leave the delayed crit, so we don't rob the player of a crit).
    for i, v in ipairs(self.level.delayed_crit) do
        if v == e then
            -- FIXME: This is awkward, since the check isn't done until crits are partially evaluated
            e:setEffect(e.EFF_OFF_BALANCE, 1, {})
            --table.remove(self.level.delayed_crit, i)
            return
        end
    end

    table.insert(self.level.delayed_crit, e)
end

---Applies any off balance effects.
-- We do this as a separate routine to delay its effects; we want to give the
-- player a chance to choose their critical effect.
function _M:processCrit()
    if self.level then
        for i, v in ipairs(self.level.delayed_crit) do
            if not v.dead then
                v:setEffect(v.EFF_OFF_BALANCE, 1, {})
            end
        end
        self.level.delayed_crit = {}
    end
end

function _M:display(nb_keyframe)
    self.player_display:toScreen(nb_keyframe)

    -- If switching resolution, blank everything but the dialog
    if self.change_res_dialog then engine.GameTurnBased.display(self, nb_keyframe) return end

    -- Now the map, if any
    if self.level and self.level.map and self.level.map.finished then
        -- Display the map and compute FOV for the player if needed
        if self.level.map.changed then
            self.player:playerFOV()
        end

        self.level.map:display(nil, nil, nb_keyframe)

        -- Display the targeting system if active
        self.target:display()

        -- And the minimap
        self.level.map:minimapDisplay(self.w - 200, 20, util.bound(self.player.x - 25, 0, self.level.map.w - 50), util.bound(self.player.y - 25, 0, self.level.map.h - 50), 50, 50, 0.6)
    end

    -- We display the player's interface
    self.flash:toScreen(nb_keyframe)
    self.logdisplay:toScreen()
    if self.show_npc_list then
        self.npcs_display:toScreen()
    else
        self.hotkeys_display:toScreen()
    end
    if self.player then self.player.changed = false end

    engine.GameTurnBased.display(self, nb_keyframe)

    -- Tooltip is displayed over all else, even dialogs
    self:targetDisplayTooltip(self.w, self.h, false, nb_keyframe)
end

--- Setup the keybinds
function _M:setupCommands()
    -- Make targeting work
    self.normal_key = self.key
    self:targetSetupKey()

    -- One key handled for normal function
    self.key:unicodeInput(true)
    self.key:addBinds
    {
        -- Movements
        MOVE_LEFT = function() self.player:moveDir(4) end,
        MOVE_RIGHT = function() self.player:moveDir(6) end,
        MOVE_UP = function() self.player:moveDir(8) end,
        MOVE_DOWN = function() self.player:moveDir(2) end,
        MOVE_LEFT_UP = function() self.player:moveDir(7) end,
        MOVE_LEFT_DOWN = function() self.player:moveDir(1) end,
        MOVE_RIGHT_UP = function() self.player:moveDir(9) end,
        MOVE_RIGHT_DOWN = function() self.player:moveDir(3) end,
        MOVE_STAY = function() self.player:useEnergy() end,

        RUN_LEFT = function() self.player:runInit(4) end,
        RUN_RIGHT = function() self.player:runInit(6) end,
        RUN_UP = function() self.player:runInit(8) end,
        RUN_DOWN = function() self.player:runInit(2) end,
        RUN_LEFT_UP = function() self.player:runInit(7) end,
        RUN_LEFT_DOWN = function() self.player:runInit(1) end,
        RUN_RIGHT_UP = function() self.player:runInit(9) end,
        RUN_RIGHT_DOWN = function() self.player:runInit(3) end,

        -- Hotkeys
        HOTKEY_1 = function() self.player:activateHotkey(1) end,
        HOTKEY_2 = function() self.player:activateHotkey(2) end,
        HOTKEY_3 = function() self.player:activateHotkey(3) end,
        HOTKEY_4 = function() self.player:activateHotkey(4) end,
        HOTKEY_5 = function() self.player:activateHotkey(5) end,
        HOTKEY_6 = function() self.player:activateHotkey(6) end,
        HOTKEY_7 = function() self.player:activateHotkey(7) end,
        HOTKEY_8 = function() self.player:activateHotkey(8) end,
        HOTKEY_9 = function() self.player:activateHotkey(9) end,
        HOTKEY_10 = function() self.player:activateHotkey(10) end,
        HOTKEY_11 = function() self.player:activateHotkey(11) end,
        HOTKEY_12 = function() self.player:activateHotkey(12) end,
        HOTKEY_SECOND_1 = function() self.player:activateHotkey(13) end,
        HOTKEY_SECOND_2 = function() self.player:activateHotkey(14) end,
        HOTKEY_SECOND_3 = function() self.player:activateHotkey(15) end,
        HOTKEY_SECOND_4 = function() self.player:activateHotkey(16) end,
        HOTKEY_SECOND_5 = function() self.player:activateHotkey(17) end,
        HOTKEY_SECOND_6 = function() self.player:activateHotkey(18) end,
        HOTKEY_SECOND_7 = function() self.player:activateHotkey(19) end,
        HOTKEY_SECOND_8 = function() self.player:activateHotkey(20) end,
        HOTKEY_SECOND_9 = function() self.player:activateHotkey(21) end,
        HOTKEY_SECOND_10 = function() self.player:activateHotkey(22) end,
        HOTKEY_SECOND_11 = function() self.player:activateHotkey(23) end,
        HOTKEY_SECOND_12 = function() self.player:activateHotkey(24) end,
        HOTKEY_THIRD_1 = function() self.player:activateHotkey(25) end,
        HOTKEY_THIRD_2 = function() self.player:activateHotkey(26) end,
        HOTKEY_THIRD_3 = function() self.player:activateHotkey(27) end,
        HOTKEY_THIRD_4 = function() self.player:activateHotkey(28) end,
        HOTKEY_THIRD_5 = function() self.player:activateHotkey(29) end,
        HOTKEY_THIRD_6 = function() self.player:activateHotkey(30) end,
        HOTKEY_THIRD_7 = function() self.player:activateHotkey(31) end,
        HOTKEY_THIRD_8 = function() self.player:activateHotkey(32) end,
        HOTKEY_THIRD_9 = function() self.player:activateHotkey(33) end,
        HOTKEY_THIRD_10 = function() self.player:activateHotkey(34) end,
        HOTKEY_THIRD_11 = function() self.player:activateHotkey(35) end,
        HOTKEY_THIRD_12 = function() self.player:activateHotkey(36) end,
        HOTKEY_PREV_PAGE = function() self.player:prevHotkeyPage() end,
        HOTKEY_NEXT_PAGE = function() self.player:nextHotkeyPage() end,

        -- Actions
        CHANGE_LEVEL = function()
            local e = self.level.map(self.player.x, self.player.y, Map.TERRAIN)
            if not self.player:enoughEnergy() then return end
            if not e.change_level then
                self.log("There is no way out of this level here.")
            elseif not e.change_zone and e.change_level < 0 then
                self.log("The Empire's agents and minions are surely in pursuit. Backtracking would be suicide.")
            else
                self.player:talentCallbackAllOn("on_pre_change_level")
                self:changeLevel(e.change_zone and e.change_level or self.level.level + e.change_level, e.change_zone)
            end
        end,

        REST = function()
            self.player:restInit()
        end,

        USE_TALENTS = function()
            self:registerDialog(require("mod.dialogs.UseTalents").new(self.player))
        end,

        SAVE_GAME = function()
            self:saveGame()
        end,

        SHOW_CHARACTER_SHEET = function()
            self:registerDialog(require("mod.dialogs.CharacterSheet").new(self.player))
        end,

        PICKUP_FLOOR = function()
            if self.player.no_inventory_access then return end
            self.player:playerPickup()
        end,

        DROP_FLOOR = function()
            if self.player.no_inventory_access then return end
            self.player:playerDrop()
        end,

        USE_ITEM = function()
            if self.player.no_inventory_access then return end
            self.player:playerUseItem()
        end,

        SHOW_INVENTORY = function()
            if self.player.no_inventory_access then return end
            local d
            local titleupdater = self.player:getEncumberTitleUpdater("Inventory")

            d = self.player:showEquipInven("Inventory", nil, function(o, inven, item, button, event)
                if not o then return end
                local ud = require("mod.dialogs.UseItemDialog").new(event == "button", self.player, o, item, inven, function(_, _, _, stop)
                    d:generate()
                    d:generateList()
                    d:updateTitle(titleupdater())

                    -- In case player added or removed a light source.
                    self.player:doFOV()

                    if stop then self:unregisterDialog(d) end
                end)
                self:registerDialog(ud)
            end)
        end,

        -- Exit the game
        QUIT_GAME = function()
            self:onQuit()
        end,

        SCREENSHOT = function() self:saveScreenshot() end,

        EXIT = function()
            local menu menu = require("engine.dialogs.GameMenu").new{
                "resume",
                { "Inventory", function() self:unregisterDialog(menu) self.key:triggerVirtual("SHOW_INVENTORY") end },
                { "Character Sheet", function() self:unregisterDialog(menu) self.key:triggerVirtual("SHOW_CHARACTER_SHEET") end },
                --{ "Show Techniques Learned", function() self:unregisterDialog(menu) self:registerDialog(require("mod.dialogs.ShowTechniquesLearned").new()) end },  TODO
                "keybinds",
                "video",
                "save",
                "quit",
                "exit"
            }
            self:registerDialog(menu)
        end,

        -- Lua console
        LUA_CONSOLE = function()
            if config.settings.cheat then
                self:registerDialog(require("engine.DebugConsole").new())
            end
        end,

        -- Debug menu
        DEBUG_MODE = function()
            if config.settings.cheat then
                self:registerDialog(require("mod.dialogs.debug.DebugMenu").new())
            end
        end,

        -- Toggle monster list
        TOGGLE_NPC_LIST = function()
            self.show_npc_list = not self.show_npc_list
            self.player.changed = true
        end,

        TACTICAL_DISPLAY = function()
            if Map.view_faction then
                self.always_target = nil
                Map:setViewerFaction(nil)
            else
                self.always_target = true
                Map:setViewerFaction("players")
            end
        end,

        LOOK_AROUND = function()
            self.flash:empty(true)
            self.flash(self.flash.GOOD, "Looking around... (direction keys to select interesting things, shift+direction keys to move freely)")
            local co = coroutine.create(function()
                local x, y = self.player:getTarget{type="hit", no_restrict=true, range=2000}
                if x and y then
                    local tmx, tmy = self.level.map:getTileToScreen(x, y)
                    self:registerDialog(MapMenu.new(tmx, tmy, x, y))
                end
            end)
            local ok, err = coroutine.resume(co)
            if not ok and err then print(debug.traceback(co)) error(err) end
        end,
    }
    self.key:setCurrent()
end

function _M:setupMouse(reset)
    if reset then self.mouse:reset() end
    self.mouse:registerZone(Map.display_x, Map.display_y, Map.viewport.width, Map.viewport.height, function(button, mx, my, xrel, yrel, bx, by, event)
        -- Handle targeting
        if self:targetMouse(button, mx, my, xrel, yrel, event) then return end

        -- Handle right click menu
        if button == "right" and event == "button" and not xrel and not yrel then
            self:mouseRightClick(mx, my)
        end

        -- Handle the mouse movement/scrolling
        self.player:mouseHandleDefault(self.key, self.key == self.normal_key, button, mx, my, xrel, yrel, event)
    end)

    -- Scroll message log
    self.mouse:registerZone(self.logdisplay.display_x, self.logdisplay.display_y, self.w, self.h, function(button)
        if button == "wheelup" then self.logdisplay:scrollUp(1) end
        if button == "wheeldown" then self.logdisplay:scrollUp(-1) end
    end, {button=true})

    -- Use hotkeys with mouse
    self.mouse:registerZone(self.hotkeys_display.display_x, self.hotkeys_display.display_y, self.w, self.h, function(button, mx, my, xrel, yrel, bx, by, event)
        self.hotkeys_display:onMouse(button, mx, my, event == "button", function(text) self.tooltip:displayAtMap(nil, nil, self.w, self.h, tostring(text)) end)
    end)
    self.mouse:setCurrent()
end

--- Right mouse click on the map
function _M:mouseRightClick(mx, my, extra)
    if not self.level then return end
    local tmx, tmy = self.level.map:getMouseTile(mx, my)
    self:registerDialog(MapMenu.new(mx, my, tmx, tmy))
end

-- Quit to main menu
function _M:onQuit()
    self.player:runStop("quitting")
    self.player:restStop("quitting")

    if not self.quit_dialog and not self.player.dead and not self:hasDialogUp() then
        self.quit_dialog = Dialog:yesnoPopup("Save and exit to the menu?", "Save and exit to the menu?", function(ok)
            if ok then
                self:saveGame()
                util.showMainMenu()
            end
            self.quit_dialog = nil
        end)
    end
end

-- Exit game
function _M:onExit()
    self.player:runStop("quitting")
    self.player:restStop("quitting")

    if not self.quit_dialog and not self.player.dead and not self:hasDialogUp() then
        self.quit_dialog = Dialog:yesnoPopup("Save and exit game?", "Save and exit game?", function(ok)
            if ok then
                -- savefile_pipe is created as a global by the engine
                self:saveGame()
                savefile_pipe:forceWait()
                os.exit()
            end
            self.quit_dialog = nil
        end)
    end
end

--- Requests the game to save
function _M:saveGame()
    -- savefile_pipe is created as a global by the engine
    savefile_pipe:push(self.save_name, "game", self)
    self.log("Saving game...")
end

function _M:onRegisterDialog(d)
    -- Hide any active tooltips, using ToME's approach.
    -- Alternate approach:
    --if game.tooltip then game.tooltip:erase() end
    self.tooltip_x, self.tooltip_y = nil, nil
end

function _M:onUnregisterDialog(d)
    -- Hide any active tooltips.
    self.tooltip_x, self.tooltip_y = nil, nil
end

